### PR TITLE
Change the complex simple_commands to @datasource

### PR DESCRIPTION
**Pull request by xiangce**

### CODE REVIEW

**Comment by skontar** [comment]
I see a bunch of new Pytrhon code here, but no tests for it. I am not a big fan of parsing `ps` yet again, and hoping it is right. Do I understand correctly, that this is a temporary solution?

File: insights/specs.py [hunk 1]
```diff
@@ -176,7 +176,37 @@ def docker_container_ids(broker):
 httpd_conf = sf.glob_file(["/etc/httpd/conf/httpd.conf", "/etc/httpd/conf.d/*.conf"], name="httpd_conf"),
 httpd_conf_sos = sf.glob_file(["/conf/httpd/conf/httpd.conf", "/conf/httpd/conf.d/*.conf"], name="httpd_conf_sos", context=HostArchiveContext)
 httpd_error_log = sf.simple_file("var/log/httpd/error_log", name="httpd_error_log")
-httpd_pid = sf.simple_command("/bin/ps aux | grep /usr/sbin/httpd | grep -v grep | head -1 | awk '{print $2}'", name="httpd_pid")
+
```
Comment (skontar) [inline insights/specs.py]: Replace all this with the following:

`httpd_pid = sf.simple_command("/bin/pgrep --oldest httpd", name="httpd_pid")`

Comment (xiangce) [reply insights/specs.py]: `pgrep` looks better, I will modify it in next commit. Thank you @skontar .

**Comment by xiangce** [comment]
@skontar - No, I don't how to test the `insights/specs.py`. I think @csams or @kblake2016 will tell us how to test it before the migration.

**Comment by zhangyi733** [comment]
@csams Do we have a method/env to test if the defined datasource can really work as what we expect?

**Comment by zhangyi733** [comment]
The code logic looks good to me. But my concern is whether these commands can really work. Glad to hear other guys' thoughts.

**Comment by csams** [comment]
@zhangyi733 Hey, you can test specs on your local environment if the commands or files they rely on are available. For example, in ipython:

```python
In [1]: from insights import run
In [2]: from insights.specs import httpd_pid

In [3]: result = run(httpd_pid, print_summary=True)

INFO:insights.core.dr:Trying insights.specs.ps_lines
INFO:insights.core.dr:Trying insights.specs.httpd_pid

datasource instances:

In [4]: result.exceptions
Out[4]: 
defaultdict(list,
            {<function insights.specs.ps_lines>: [TypeError("unhashable type: 'dict'")]})
In [5]: for e, t in result.tracebacks.items():
    ...:    print(t)
    ...:     
Traceback (most recent call last):
  File "insights/core/dr.py", line 639, in run
    result = DELEGATES[component](broker)
  File "insights/core/dr.py", line 598, in __f
    return executor(func, broker, requires, optional)
  File "insights/core/dr.py", line 521, in broker_executor
    return func(broker)
  File "insights/specs.py", line 195, in ps_lines
    "COMMAND": p_splits[-1]
TypeError: unhashable type: 'dict'
```
It looks like a helper function is trying to store a dictionary in a set, which isn't allowed because dicts aren't hashable.

**Comment by xiangce** [comment]
@csams - thanks for your sample code, I will look into it.

**Comment by xiangce** [comment]
@skontar @zhangyi733 @csams @bfahr 

Hey guys,

I have tested this PR on an RHEL7 satellite 6 Server, it works fine:

~~~python
In [1]: from insights.specs import httpd_limits, httpd_pid, package_provides_java, ipcs_s_i, tomcat_vdc_targeted                                                                               
                                                                                                                                                                                               
In [2]: from insights import run                                                                                                                                                              
                                                                                                                                                                                              
In [3]: ret = run(ipcs_s_i, print_summary=True)                                                
INFO:insights.core.dr:Trying insights.specs.ipcs_s                                                                                                                                            
INFO:insights.core.dr:Trying insights.specs.semid                                                                                                                                             
INFO:insights.core.dr:Trying insights.specs.ipcs_s_i                                           
                                                                                                                                                                                              
                                                                                                                                                                                              
datasource instances:                                                                          
'insights.specs.ipcs_s:'                                                                       
CommandOutputProvider("/usr/bin/ipcs -s")                                                                                                                                                     
                                                                                                                                                                                               
'insights.specs.ipcs_s_i:'                                                                                                                                                                     
[CommandOutputProvider("/usr/bin/ipcs -s -i 557058"),                                                                                                                                         
 CommandOutputProvider("/usr/bin/ipcs -s -i 589827"),                                                                                                                                         
 CommandOutputProvider("/usr/bin/ipcs -s -i 786441"),                                          
 CommandOutputProvider("/usr/bin/ipcs -s -i 1343496"),                                                
 CommandOutputProvider("/usr/bin/ipcs -s -i 1310727"),
 CommandOutputProvider("/usr/bin/ipcs -s -i 491520"),
 CommandOutputProvider("/usr/bin/ipcs -s -i 688134"),
 CommandOutputProvider("/usr/bin/ipcs -s -i 655365"),
 CommandOutputProvider("/usr/bin/ipcs -s -i 1441804"),
 CommandOutputProvider("/usr/bin/ipcs -s -i 622596"),
 CommandOutputProvider("/usr/bin/ipcs -s -i 524289"),
 CommandOutputProvider("/usr/bin/ipcs -s -i 1376266"),
 CommandOutputProvider("/usr/bin/ipcs -s -i 1409035")]

'insights.specs.semid:'
set(['1310727',
     '1343496',
     '1376266',
     '1409035',
     '1441804',
     '491520',
     '524289',
     '557058',
     '589827',
     '622596',
     '655365',
     '688134',
     '786441'])

In [5]: ret = run(httpd_limits, print_summary=True)
INFO:insights.core.dr:Trying insights.specs.httpd_pid
INFO:insights.core.dr:Trying insights.specs.httpd_limits


datasource instances:
'insights.specs.httpd_limits:'
[CommandOutputProvider("/bin/cat /proc/5920/limits")]

'insights.specs.httpd_pid:'
CommandOutputProvider("/usr/bin/pgrep -o httpd")

In [6]: ret = run(package_provides_java, print_summary=True)
INFO:insights.core.dr:Trying insights.specs.ps_lines
INFO:insights.core.dr:Trying insights.specs.package_and_java
INFO:insights.core.dr:Trying insights.specs.package_provides_java


datasource instances:
'insights.specs.package_and_java:'
set(['java java-1.7.0-openjdk-headless-1.7.0.141-2.6.10.5.el7.x86_64'])

'insights.specs.package_provides_java:'
[CommandOutputProvider("echo java java-1.7.0-openjdk-headless-1.7.0.141-2.6.10.5.el7.x86_64")]

'insights.specs.ps_lines:'
[{'COMMAND': '/usr/lib/systemd/systemd --switched-root --system --deserialize 21',
...

In [7]: ret = run(tomcat_vdc_targeted, print_summary=True)
INFO:insights.core.dr:Trying insights.specs.ps_lines
INFO:insights.core.dr:Trying insights.specs.tomcat_home
INFO:insights.core.dr:Trying insights.specs.tomcat_vdc_targeted


datasource instances:
'insights.specs.ps_lines:'
[{'COMMAND': '/usr/lib/systemd/systemd --switched-root --system --deserialize 21',
  'PID': '1'},
...

'insights.specs.tomcat_home:'
set(['/usr/share/tomcat'])

'insights.specs.tomcat_vdc_targeted:'
[CommandOutputProvider("/bin/grep -R -s 'VirtualDirContext' --include '*.xml' /usr/share/tomcat")]

~~~

**Comment by xiangce** [comment]
Here is the test result on RHEL6:

```python
In [1]: from insights.specs import httpd_limits, httpd_pid, package_provides_java, ipcs_s_i, tomcat_vdc_targeted                                              
                                                                                               
In [2]: from insights import run                                                                                                                                                               
                                                                                                                                                                                               
In [3]: ret = run(ipcs_s_i, print_summary=True)                                                                                                                                                
INFO:insights.core.dr:Trying insights.specs.ipcs_s                                                                                                                                             
INFO:insights.core.dr:Trying insights.specs.semid                                                                                                                                              
INFO:insights.core.dr:Trying insights.specs.ipcs_s_i                                                                                                                                           
                                                                                                                                                                                               
                                                                                                                                                                                               
datasource instances:                                                                                                                                                                          
'insights.specs.ipcs_s:'                                                                                                                                                                       
CommandOutputProvider("/usr/bin/ipcs -s")                                                                                                                                                      
                                                                                                                                                                                               
'insights.specs.ipcs_s_i:'                                                                                                                                                                     
[CommandOutputProvider("/usr/bin/ipcs -s -i 32769"),                                                                                                                                           
 CommandOutputProvider("/usr/bin/ipcs -s -i 262152"),
 CommandOutputProvider("/usr/bin/ipcs -s -i 1114146"),
...
'insights.specs.semid:'
set(['0',
     '1015839',
     '1048608',
     '1081377',
     '1114146',
     '1146915'
...
In [4]: ret = run(httpd_limits, print_summary=True)
INFO:insights.core.dr:Trying insights.specs.httpd_pid
INFO:insights.core.dr:Trying insights.specs.httpd_limits


datasource instances:
'insights.specs.httpd_limits:'
[CommandOutputProvider("/bin/cat /proc/23844/limits")]

'insights.specs.httpd_pid:'
CommandOutputProvider("/usr/bin/pgrep -o httpd")


In [5]: ret = run(package_provides_java, print_summary=True)
INFO:insights.core.dr:Trying insights.specs.ps_lines
INFO:insights.core.dr:Trying insights.specs.package_and_java
INFO:insights.core.dr:Trying insights.specs.package_provides_java


datasource instances:
'insights.specs.package_and_java:'
set(['/usr/bin/java java-1.8.0-ibm-1.8.0.4.1-1jpp.1.el6_8.x86_64',
     '/usr/lib/jvm/java/bin/java java-1.8.0-ibm-1.8.0.4.1-1jpp.1.el6_8.x86_64'])

'insights.specs.package_provides_java:'
[CommandOutputProvider("echo /usr/lib/jvm/java/bin/java java-1.8.0-ibm-1.8.0.4.1-1jpp.1.el6_8.x86_64"),
 CommandOutputProvider("echo /usr/bin/java java-1.8.0-ibm-1.8.0.4.1-1jpp.1.el6_8.x86_64")]

'insights.specs.ps_lines:'
[{'COMMAND': '/sbin/init', 'PID': '1'},
...
 {'COMMAND': '/usr/sbin/httpd.worker', 'PID': '23844'}]


In [6]: ret = run(tomcat_vdc_targeted, print_summary=True)
INFO:insights.core.dr:Trying insights.specs.ps_lines
INFO:insights.core.dr:Trying insights.specs.tomcat_home
INFO:insights.core.dr:Trying insights.specs.tomcat_vdc_targeted


datasource instances:
'insights.specs.ps_lines:'
...
'insights.specs.tomcat_home:'
set(['/usr/share/tomcat6'])
```

File: insights/specs.py (hunk 1 above)
Comment (xiangce) [reply insights/specs.py]: @skontar - I have updated and tested the related datasources. Please help check if they are OK. Thanks.

**Comment by csams** [comment]
Is 0 a valid value in `insights.specs.semid`?

**Comment by xiangce** [comment]
AFAIK it's a good `semid`, here is the output on my satellite5_rhel6:

~~~
[root@liuxc-rhel6-sat5 ~]# ipcs -s | head

------ Semaphore Arrays --------
key        semid      owner      perms      nsems     
0x00000000 0          root       600        1         
0x00000000 32769      root       600        1         
0x0052e2c1 65538      postgres   600        17        
0x0052e2c2 98307      postgres   600        17        
0x0052e2c3 131076     postgres   600        17        
0x0052e2c4 163845     postgres   600        17        
0x0052e2c5 196614     postgres   600        17        
[root@liuxc-rhel6-sat5 ~]# ipcs -s -i 0

Semaphore Array semid=0
uid=0    gid=0   cuid=0  cgid=0
mode=0600, access_perms=0600
nsems = 1
otime = Mon Nov 27 14:12:40 2017  
ctime = Mon Nov 27 14:12:39 2017  
semnum     value      ncount     zcount     pid       
0          1          0          0          155
~~~

**Comment by xiangce** [comment]
Will separate into individual PRs for each datasource.

**Comment by xiangce** [comment]
For `tomcat_home`, please see #832

**Comment by xiangce** [comment]
For "httpd_V", see #833

**Pull request closed without merging**